### ISSUE TITLE
IPv6 connection to TVHeadend doesn't seem to work with Matrix/pvr.hts 8.1

### ISSUE THREAD

**Issue by mamarley** [kodi-pvr/pvr.hts#483]
I just upgraded my Kodi installation to the newly-released Matrix beta (and pvr.hts 8.1 from the add-on repository).  Now that I have done this, pvr.hts will only connect to TVHeadend using IPv4.  Previously (with Kodi 18.x) it would use IPv6 if it was available.  Now with a domain name that has both A and AAAA it will use IPv4 and with a domain name that has only AAAA it refuses to connect entirely.  I don't have an advancedsettings.xml.  I am running Windows 10 19042.630 on the client.

**Comment by Jalle19** [comment]
Are you sure tvheadend is listening on IPv6? It's not enabled by default.

**Comment by mamarley** [comment]
Quite sure.  pvr.hts was using IPv6 successfully with Kodi 18.  I also telnetted the TVH server's HTTP port on IPv6 to make sure nothing had changed.

**Comment by Jalle19** [comment]
Can you check if Kodi is using IPv6 for anything else, i.e. is this limited to just pvr.hts or Kodi as a whole?

**Comment by FernetMenta** [comment]
the introduction of kissnet broke it https://github.com/kodi-pvr/pvr.hts/pull/475

**Comment by mamarley** [comment]
And it appears, according to the comments in https://github.com/Ybalrid/kissnet/issues/23, that kissnet uses IPv4 unless the caller specifically requests IPv6.  That seems silly because it forces the logic to determine whether or not IPv6 is even available, attempt a connection, and potentially fall back to IPv4 into the client, when ideally for IPv6 adoption it should be automatic and transparent.

**Comment by ksooo** [comment]
I'm sure the kissnet author is open for PRs. My PRs got merged very fast...

**Comment by mamarley** [comment]
Thanks for putting that together!  I had just been looking over the kissnet code myself and I was wondering why it seemed to needlessly differentiate between IPv4 and IPv6 sockets instead of using AF_UNSPEC, but I'm far from an expert on low-level socket programming, so I wasn't confident that I had the right idea.

**Comment by mamarley** [comment]
This doesn't seem to be quite fixed.  Now the connection just hangs if I use a dual-stack hostname or IPv6 address.  If I use an IPv4 address, it still works.  I'm going to sniff the traffic and see if I can figure out what is going on.

**Comment by mamarley** [comment]
The packet sniffing reveals that when trying to connect to a dual-stack hostname or an IPv6 address, no traffic ever leaves the client system on port 9982.  In the case of the hostname, it does get as far as the DNS lookup, which is done for both A and AAAA.  I have attached the "kodi.log" with debug logging and component-specific logging enabled, but it doesn't look like anything useful is in there.
[kodi.log](https://github.com/kodi-pvr/pvr.hts/files/5608025/kodi.log)

**Comment by mamarley** [comment]
I was also able to reproduce using a minimal test case (on Linux, where my Kodi client is on Windows, so this isn't platform specific):

```
#include <iostream>
#include <thread>
#include <chrono>

#include "kissnet.hpp"
using namespace std::chrono_literals;

int main(){
	kissnet::tcp_socket a_socket(kissnet::endpoint("<ADDRESS_OR_HOSTNAME>:9982"));
	a_socket.connect();
	std::this_thread::sleep_for(15s);
}
```

If I put in the hostname of my server, I never see any network activity (besides the DNS lookup).  If I put in the IPv4 IP, I can see the connection in TVHeadend for 15 seconds until the example times out and quits.

**Comment by mamarley** [comment]
I don't think GitHub sends notifications for mentions, so I will summarize what I found in a comment as well.  It appears that kissnet wasn't using the correct lengths when memcpy()ing the getaddrinfo() results and calling the connect() function, which was truncating IPv6 addresses and causing connections to fail.  I opened a PR to fix this issue.

I don't have the proper setup to compile pvr.hts for Windows in order to test it directly, but my PR does allow for the minimal testcase I posted above to successfully make IPv6 connections now.
